### PR TITLE
[MIG] Migrate currency_rate_update to V8.

### PR DESCRIPTION
**Pull request by feketemihai**

### CODE REVIEW

**Comment by feketemihai** [comment]
@nbessi @guewen Is this supposed to happen? Tests from OCA/OCB failed in openerp.addons.base.tests.test_xmlrpc???

**Comment by nbessi** [comment]
@feketemihai yes they fail if your are in multi process there is an issue about this.

**Comment by feketemihai** [comment]
@nbessi And what do i have to do now to restart the coverage?

**Comment by feketemihai** [comment]
@guewen @yvaucher I have included all the comments from pull 70 to this one...i started from scratch with git mv...can you review it to see if everything fits. Thanks.

.

**Comment by guewen** [comment]
Is it normal that you added commits on account_journal_period_close?

**Comment by feketemihai** [comment]
I wanted to fix the pylint errors from this module, and i removed them at the end.

**Comment by guewen** [comment]
Ok I see. Do you know you can use `git rebase -i [sha_of_starting_commit]` to rewrite the history and remove the useless commits? (after a rewrite of the history, you will need to `git push -f`).

File: __unported__/account_reversal/__init__.py [hunk 1]
```diff
@@ -20,5 +20,5 @@
 #
 ##############################################################################
 
-from . import account_reversal
-from . import wizard
+import account_reversal
+import wizard
```
Comment (guewen) [inline __unported__/account_reversal/__init__.py]: The imports were correct there.

Comment (guewen) [inline __unported__/account_reversal/__init__.py]: Oh, it is in `account_reversal`. Could you rebase from 8.0?

File: currency_rate_update/model/currency_rate_update.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        """Refresh the currencies rates !!for all companies now"""
+        factory = Currency_getter_factory()
+        curr_obj = self.env['res.currency']
+        rate_obj = self.env['res.currency.rate']
+        company = self.company_id
+        # The multi company currency can be set or no so we handle
+        # The two case
+        if company.auto_currency_up:
+            main_currencies = curr_obj.search(
+                [('base', '=', True), ('company_id', '=', company.id)])
```
Comment (guewen) [inline currency_rate_update/model/currency_rate_update.py]: You could add a limit=1 in the search because only the first will be used

File: currency_rate_update/model/currency_rate_update.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+            if not main_currencies:
+                # If we can not find a base currency for this company
+                # we look for one with no company set
+                main_currencies = curr_obj.search(
+                    [('base', '=', True), ('company_id', '=', False)])
+            if main_currencies:
+                main_curr = main_currencies[0]
```
Comment (guewen) [inline currency_rate_update/model/currency_rate_update.py]: `[0]` is not needed because a `recordset` with 1 item will act like a record (direct access to the attributes of the first record)

File: currency_rate_update/model/currency_rate_update.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+    @api.multi
+    def run_currency_update(self):
+        "Update currency at the given frequence"
+        ctx = dict(self._context)
+        current_date = fields.Date.today()
+        services = self.search([('next_run', '=', current_date)])
+        ctx['cron'] = True
+        for service in services:
+            service.with_context(ctx).refresh_currency()
```
Comment (guewen) [inline currency_rate_update/model/currency_rate_update.py]: the better is 

```
service.with_context(cron=True).refresh_currency()
```

so you can remove the lines 302 and 305
Also, you don't need to loop over the services you can do:

```
services.with_context(cron=True).refresh_currency()
```

**Comment by guewen** [comment]
Needs a rebase but seems good otherwise.
Thanks

**Comment by feketemihai** [comment]
@guewen i made "git reset --hard 97878c2" but still didn't erase the history commits...Sorry, i had to go with 1 commit before...Thanks.

**Comment by feketemihai** [comment]
@guewen Can you tell me how it's the best way to update translation...for every language already translated...If i will export the pot file again the lines won't be the same...

**Comment by yvaucher** [comment]
@feketemihai You can override completely po[t] files. Translations should be correctly sorted now. When this module was developed it wasn't. https://bugs.launchpad.net/ocb-server/+bug/971412

**Comment by feketemihai** [comment]
@yvaucher Yannick, when i export the pot file, the field multi_company_currency_enable from res.company is not translated (the help either), only if i store it, but if i store it will not be recomputed, as far as i know. Even if i put translate=True, is not taken in consideration. It's ok if i put the strings directly in the pot file?

**Comment by feketemihai** [comment]
@yvaucher It's a new API bug, because i've checked in event module the field count_registrations, which has a compute method and don't appear as well in the pot file exported. I will leave like this and make an issue on ODOO. Thanks.

**Comment by yvaucher** [comment]
@feketemihai Can you point to the Odoo issue when done ? You can still add the string by hand in pot file. It will be ok.

Or you could make that field stored while generating the pot file so you get all translation. Without commiting the stored option change.

Thanks

**Comment by yvaucher** [comment]
@feketemihai seeing your last commit, please prefer a rebase:

`git rebase -i origin/8.0`

**Comment by feketemihai** [comment]
@yvaucher Issue for translation of computed fields is https://github.com/odoo/odoo/issues/3745. And for readme.md file, description not present in the pot file is https://github.com/odoo/odoo/issues/3747.

**Comment by feketemihai** [comment]
@guewen @yvaucher Can you have a final review on this...

**Comment by guewen** [comment]
Thanks @feketemihai 

:+1:

File: currency_rate_update/README.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
+Import exchange rates from the Internet.
```
Comment (pedrobaeza) [inline currency_rate_update/README.md:1]: This must be `README.rst`, not md, because Odoo is not going to recognize it.

File: currency_rate_update/README.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+The update can be set under the company form.
+You can set for each services which currency you want to update.
+The logs of the update are visible under the service note.
+You can active or deactivate the update.
+The module uses internal ir-cron feature from OpenERP, so the job is
+launched once the server starts if the 'first execute date' is before
+the current day.
+The module supports multi-company currency in two ways:
+
+*    the currencies are shared, you can set currency update only on one
```
Comment (pedrobaeza) [inline currency_rate_update/README.md:45]: This is not indented the same for the two lines, and RST parser is not going to deal with it.

File: currency_rate_update/model/company.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            company.multi_company_currency_enable = \
+                1 if company_currency else 0
+
+    @api.one
+    def button_refresh_currency(self):
+        """Refresh the currencies rates !!for all companies now"""
+        self.services_to_use.refresh_currency()
+
+    _inherit = "res.company"
+    _name = "res.company"
```
Comment (pedrobaeza) [inline currency_rate_update/model/company.py]: You don't need to put name if you inherit and the name remains the same.

File: currency_rate_update/model/currency_rate_update.py [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+    @api.onchange('service')
+    def _onchange_service(self):
+        currency_list = ''
+        if self.service:
+            currencies = []
+            currency_list = supported_currency_array
+            company_id = False
+            if self.company_id.multi_company_currency_enable:
+                company_id = self.company_id.id
+            if self.service == 'ECB_getter':
```
Comment (pedrobaeza) [inline currency_rate_update/model/currency_rate_update.py]: Why don't you put this as a mapping dictionary at the begginning? This way, you don't need the sequence of 'if'

File: currency_rate_update/services/__init__.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Comment (pedrobaeza) [inline currency_rate_update/services/__init__.py:20]: Is this file needed if you remove 'services' from root init? And does import work as expected for the other classes?

**Comment by pedrobaeza** [comment]
Some remarks inline and one last question: have you updated translation template?

File: currency_rate_update/README.md (hunk 5 above)
Comment (yvaucher) [inline currency_rate_update/README.md:1]: @pedrobaeza https://github.com/odoo/odoo/blob/8.0/openerp/modules/module.py#L42

It should detect README.rst, README.txt and README.md

Comment (feketemihai) [inline currency_rate_update/README.md:1]: As a description it's detected, but it's not added in the translation terms.

File: currency_rate_update/model/currency_rate_update.py (hunk 8 above)
Comment (feketemihai) [inline currency_rate_update/model/currency_rate_update.py]: It's ok to do it dinamically..like in services/currency_getter line 76, there i've done it because otherwise pylint would generate error of having imports and not used them.

File: currency_rate_update/services/__init__.py (hunk 9 above)
Comment (feketemihai) [inline currency_rate_update/services/__init__.py:20]: I tested without it, and having errors about the currency_getter...

File: currency_rate_update/README.md (hunk 6 above)
Comment (feketemihai) [inline currency_rate_update/README.md:45]: Basically it's a indented paragraph describing the 2 ways...it was the same for older descriptions in openerp.py

**Comment by feketemihai** [comment]
@pedrobaeza Yes, i've updated the template...and made 2 issues on Odoo...Issue for translation of computed fields is odoo/odoo#3745. And for readme.md file, description not present in the pot file is odoo/odoo#3747.

File: currency_rate_update/README.md (hunk 5 above)
Comment (pedrobaeza) [inline currency_rate_update/README.md:1]: Yeah, but sintax is parsed in Github as Markdown, not RST, provoking inconsistencies, because expected syntax in Odoo is RST, or this has been also improved?

File: currency_rate_update/model/currency_rate_update.py (hunk 8 above)
Comment (pedrobaeza) [inline currency_rate_update/model/currency_rate_update.py]: I'm talking about having:

```
services_mapping = {
    'ECB_getter': ECB_supported_currency_array,
    'RO_BNR_getter': RO_BNR_supported_currency_array,
...
}
```

And here use:

```
currency_list = services_mapping[self.service]
```

File: currency_rate_update/README.md (hunk 6 above)
Comment (pedrobaeza) [inline currency_rate_update/README.md:45]: Yeah, but old description is wrong. You should align second line with the first not token character (in this case, the 't' of 'the'). Now it is one space minus.

File: currency_rate_update/model/currency_rate_update.py (hunk 8 above)
Comment (feketemihai) [inline currency_rate_update/model/currency_rate_update.py]: Yes but i didn't found for Yahoo how to get all the supported currencies..that's why i've putted the "if"s...

**Comment by pedrobaeza** [comment]
About the second one, I can tell you that this is an old discussion. Module descriptions are translated and put it on base namespace, due to technical restrictions. That's why you can't have custom modules with their description translated. This has difficult solution.

File: currency_rate_update/model/currency_rate_update.py (hunk 8 above)
Comment (pedrobaeza) [inline currency_rate_update/model/currency_rate_update.py]: But this is because YAHOO_getter is not completed and doesn't report supported currencies, doesn't it?

Comment (feketemihai) [inline currency_rate_update/model/currency_rate_update.py]: It's about the methid, here is from_currency  - to_currency and i didn't found anything on their site...i just found something (http://finance.yahoo.com/webservice/v1/symbols/allcurrencies/quote?format=json) i will process the response and to add them in a variable..i hope they are all...

Comment (pedrobaeza) [inline currency_rate_update/model/currency_rate_update.py]: Thanks for the effort you are doing!

**Comment by pedrobaeza** [comment]
Thanks for the changes.

:+1:

**Comment by pedrobaeza** [comment]
I proceed with the merge.

**Comment by feketemihai** [comment]
Thanks for the merge...

**Comment by yvaucher** [comment]
@feketemihai Thanks you for your contribution. This is one of the most downloaded module of OCA in 7.0. Many will be glad it finally landed on 8.0.
